dask/dataframe/groupby.py

```diff
@@ -132,6 +136,243 @@ def _nunique_series_aggregate(df):
     return df.groupby(df.columns[1])[df.columns[0]].nunique()
 
 
+###############################################################
+# Aggregate support
+#
+# Aggregate is implemented as:
+#
+# 1. group-by-aggregate all partitions into intermediate values
+# 2. collect all partitions into a single partition
+# 3. group-by-aggregate the result into intermediate values
+# 4. transform all intermediate values into the result
+#
+# In Step 1 and 3 the dataframe is grouped on the same columns.
+#
+###############################################################
+
+def _aggregate_meta(df, index, levels, chunk_funcs, agg_funcs, finalizers):
+    stage1 = _groupby_apply_funcs(df, funcs=chunk_funcs, by=index)
+    stage2 = _groupby_apply_funcs(stage1, funcs=agg_funcs, level=levels)
+    return _agg_finalize(stage2, finalizers)
+
+
+def _normalize_spec(spec, non_group_columns):
+    """
+    Return a list of ``(result_column, func, input_column)`` tuples.
+    """
+    if not isinstance(spec, dict):
+        spec = collections.OrderedDict(zip(non_group_columns, it.repeat(spec)))
+
+    res = []
+
+    if isinstance(spec, dict):
+        for input_column, subspec in spec.items():
+            if isinstance(subspec, dict):
+                res.extend(((input_column, result_column), func, input_column)
+                           for result_column, func in subspec.items())
+
+            else:
+                if not isinstance(subspec, list):
+                    subspec = [subspec]
+
+                res.extend(((input_column, funcname(func)), func, input_column)
+                           for func in subspec)
+
+    else:
+        raise ValueError("unsupported agg spec of type {}".format(type(spec)))
+
+    compounds = (list, tuple, dict)
+    use_flat_columns = not any(isinstance(subspec, compounds)
+                               for subspec in spec.values())
+
+    if use_flat_columns:
+        res = [(input_col, func, input_col) for (_, func, input_col) in res]
+
+    return res
+
+
+def _build_agg_args(spec):
+    """
+    Create the transformation functions for a normalized aggregate spec.
+
+    Returns
+    -------
+    chunk_funcs: TODO: document
+
+    agg_funcs: TODO: document
+
+    finalizers: TODO: document
+    """
+    # generator for consecutive IDs (for intermediate results). Ids are
+    # generated for (result, func) pairs. Once an id for a pair has been
+    # generated it is reused in subsequent id requests.
+    ids = it.count()
+    # TODO: add func, input_column to intermediate name for easier debugging
+    id_map = collections.defaultdict(lambda: 'intermediate-{}'.format(next(ids)))
```

Comment:
Currently, the skipping of recomputation is implemented by a somewhat arcane combination of `itertools.count` and `defaultdict`. What do you think about implementing it as a separate class? This would also make it simpler to generate easy to interpret intermediate names.